tcpdump.c

```diff
@@ -0,0 +1,239 @@
+
+    mreq.mr_ifindex = if_index;
+    mreq.mr_type = PACKET_MR_PROMISC;
+
+	opt_name = set ? PACKET_ADD_MEMBERSHIP : PACKET_DROP_MEMBERSHIP;
+    CHECK(setsockopt(sock, SOL_PACKET, opt_name, &mreq, sizeof(mreq)) != -1);
+    // TODO: this is a side effect. what if i don't want to print?
+    printf("set interface: %s to promisc %d\n", if_name, set);
+
+cleanup:
+    return ret_status;
+}
+
+static error_status_t bind_raw_socket(int sock, const char* if_name) {
+    error_status_t ret_status = STATUS_SUCCESS;
+    int if_index = -1;
+    struct sockaddr_ll sll = {0};
+    
+    CHECK(if_name != NULL);
+    CHECK_FUNC(iface_name_to_index(sock, if_name, &if_index));
+
+    // to bind a raw socket to a specific interface 
+    sll.sll_family = AF_PACKET;
+    sll.sll_protocol = htons(ETH_P_ALL);
+    sll.sll_ifindex = if_index;
+
+    CHECK_STR(bind(sock, (struct sockaddr*)&sll, sizeof(sll)) >= 0, "Failed binding to interface");
+
+cleanup:
+    // TODO: i don't close here even if it fails because i want main func to close it? 
+    //close(sock);
+    return ret_status;
+
+}
+
+static error_status_t create_raw_socket(int* res) {
+    error_status_t ret_status = STATUS_SUCCESS;
+    int sock = -1;
+    
+    CHECK(res != NULL);
+
+    sock = socket(PF_PACKET, SOCK_RAW, htons(ETH_P_ALL));
+    CHECK_STR(sock >= 0, "Creating raw socket failed");
+
+    *res = sock;
+cleanup:
+    return ret_status;
+}
+static error_status_t handle_packet(unsigned char* packet) {
+    error_status_t ret_status = STATUS_SUCCESS;
+    
+    CHECK(packet != NULL);
+
+    // Basic packet parser
+    // add more: ip level, tcp/udp, blah blah
+    struct ethhdr* eth_h = (struct ethhdr*)packet;
+	printf("Ethernet Header\n");
+	printf("\t|-Source Address : %.2X:%.2X:%.2X:%.2X:%.2X:%.2X\n", eth_h->h_source[0],eth_h->h_source[1],eth_h->h_source[2],
+   		eth_h->h_source[3],eth_h->h_source[4],eth_h->h_source[5]);
+	printf("\t|-Destination Address : %.2X:%.2X:%.2X:%.2X:%.2X:%.2X\n", eth_h->h_dest[0],eth_h->h_dest[1],eth_h->h_dest[2],
+   		eth_h->h_dest[3],eth_h->h_dest[4],eth_h->h_dest[5]);
+
+cleanup:
+    return ret_status;
+}
+
+static error_status_t sniff(int sock, int fd) {
+    error_status_t ret_status = STATUS_SUCCESS;
+    struct sockaddr_ll sll = {0};
+    socklen_t sll_len = sizeof(sll);
+    unsigned char buffer[MAX_PKT_LEN] = {0};
+    ssize_t r_bytes = 0;
+
+    while (running) {
+        // 0 is for the flags field, no special flags are used
+        r_bytes = recvfrom(sock, buffer, sizeof(buffer), 0, (struct sockaddr*)&sll,
+                           &sll_len);
+        if (r_bytes == MAX_PKT_LEN) {
+            printf("Packet len too long\n");
+            continue;
+        }
+
+        // TODO: is this the best way to handle this?
+        // the problem here is that when we SIGINT recv terminates,
+        // we want to handle this without CHECKing as it goes into cleanup.
+        if (r_bytes == -1) {
```

Comment:
Notice that if `errno != EINTR` you just continue as usual. I would do:
```c
if (r_bytes == -1) {
    CHECK(errno == EINTR);
}
```
Also, EINTR isn't the only ERRNO that's OK to get here, read the recvfrom manpage.